### PR TITLE
Make end-to-end Tests for Editable Textfield

### LINKED ISSUE #97: Subtask: Create End-to-End usability tests for `OSCALEditableTextField`

### PR DESCRIPTION
**Pull request by schen2166**
Enter:
Makes sure that any edits save after a Enter key press
Tests the title saves current title if Only Enter is Pressed

ESC:
Tests if ESC key press will cancel any edits made. reverts to original title before edit
Tests text field will remains same if only ESC is pressed.

Also changed to afterEach in each test suite to make sure each test runs with the same default parameters.

Note: When reviewed, may want to combine this cypress test with OSCAL_Edit_Tests.cy.js

Closes the Issue https://github.com/EasyDynamics/oscal-editor-deployment/issues/97

### CODE REVIEW

**Comment by tuckerzp** [comment]
A few things just to keep on track:
1. Are your local `Edit` tests still failing? They seem to be passing when run through the workflow.
2. I personally think that keeping the `TextField` tests inside the confines of `LoadTests` & `EditTests` makes the most sense. 
3. I am not sure if linting code is the best thing to be doing for this specific pr. I am not against it completely, but it makes the `diff` a bit tricky to understand what is actually being changed.

**Comment by schen2166** [comment]
Created end-to-end tests for Enter and Esc buttons
Enter:
Makes sure that any edits save after a Enter key press
Tests the title saves current title if Only Enter is Pressed

ESC:
Tests if ESC key press will cancel any edits made. reverts to original title before edit
Tests text field will remains same if only ESC is pressed.

Also changed to afterEach in each test suite to make sure each test runs with the same default parameters.
Closes the Issue #97

**Comment by tuckerzp** [comment]
Just to make reviewing this a bit easier, what are the titles of the tests that are actually being added?

**Comment by schen2166** [comment]
> Just to make reviewing this a bit easier, what are the titles of the tests that are actually being added?

There are a total of 2 test suites each with 2 tests
Suite 1 (titled Test Editing with Enter key)
Test 1: Successfully Edits Title with Enter
Test 2: Successfully keeps title on Enter key

Suite 2: (Titled Test Cancel Edit with ESC)
Test 1: Successfully cancels any changes on ESC keypress
Test 2 ESC key cancels textfield edit action

These lines of code are found from lines 139-199 on OSCAL_Edit_Test.cy.js

**Review by tuckerzp — DISMISSED** [review]
Looks good to me! Thanks for your work on this @schen2166

File: end-to-end-tests/cypress/e2e/OSCAL/OSCAL_Textfield_Tests.cy.js [hunk 1]
```diff
@@ -41,24 +38,34 @@ describe("Cancel on ESC Keypress", () => {
...
     cy.get(`[data-testid="textField-system-security-plan-metadata-title"]`)
       .click()
       .clear()
-      .type("{esc}")
-      .should("have.value", SSP_TITLE_ORIG);
+      .type("{esc}");
+      cy.get(`[aria-label="edit-system-security-plan-metadata-title"]`).click();
+      cy.get(`[data-testid="textField-system-security-plan-metadata-title"]`)
+      .click()
+      .should("have.value", SSP_ALTERNATE);
```
Comment (kylelaker) [inline end-to-end-tests/cypress/e2e/OSCAL/OSCAL_Textfield_Tests.cy.js:69]: Formatting is very weird here. The `cy.get` lines likely need to be un-indented? Same in a few other places.

**Review by laurelmay — CHANGES_REQUESTED** [review]

**Review by laurelmay — COMMENTED** [review]
These tests _really_ heavily rely on `aria-label` to select the components. Is there a better way for us to be able to find these edit buttons? Or is the `aria-label` the best option we have right now.

Usually, we want to think about accessing elements in the way that an end user would interact with the page. Even users who would be using accessibility tools likely aren't purely navigating by `aria-label`.

I assume typically a user would find the control, the component, and the text they want to edit and click the nearest "edit" button to that. Does our current DOM structure allow for that?

If not, that's okay and we can stick with the `aria-label`s for now; but it is a larger fix that we need to be aware of needing to correct in the future.

**Comment by tuckerzp** [comment]
> I believe currently, the only way to access editing on the buttons is to click on the edit button.

Would it be difficult to emulate clicking on the edit button then editing the nearest text field?

**Comment by schen2166** [comment]
Not sure what you mean by that but you can see visually below. If the window is small, the edit button would be overshadowed by other components which make in inaccessible. We could make the text box clickable to resolve this.

Image above is window in full screen with is fine. Below is the web app at a reduced size. The `Edit` button should be next to **"BASELINE"** but the window size knocks it down to the next line. Which is behind the **"Joint Task Force, Transformation Intiative"** box component which is not clickable.

File: end-to-end-tests/cypress/e2e/OSCAL/OSCAL_Edit_Tests.cy.js [hunk 2]
```diff
@@ -43,78 +43,154 @@ describe('Test Editing System Security Plan Source', () => {
...
     // Verify updated value
-    cy.contains('h2', CONTROL_NAME).should('be.visible')
-  })
-})
\ No newline at end of file
+    cy.contains("h2", CONTROL_NAME).should("be.visible");
+  });
+});
+
+describe("Test Editing Title Using Enter key", () => {
```
Comment (tuckerzp) [inline end-to-end-tests/cypress/e2e/OSCAL/OSCAL_Edit_Tests.cy.js:130]:
```suggestion
describe("Editing Title Using Enter key", () => {
```

File: end-to-end-tests/cypress/e2e/OSCAL/OSCAL_Edit_Tests.cy.js [hunk 3]
```diff
@@ -43,78 +43,154 @@ describe('Test Editing System Security Plan Source', () => {
...
+    cy.get(`[aria-label="edit-system-security-plan-metadata-title"]`).click();
+    cy.get(`[data-testid="textField-system-security-plan-metadata-title"]`)
+      .click()
+      .clear()
+      .type("{enter}");
+    cy.contains(SSP_TITLE_ORIG);
+  });
+});
+
+describe("Cancel on ESC Keypress", () => {
```
Comment (tuckerzp) [inline end-to-end-tests/cypress/e2e/OSCAL/OSCAL_Edit_Tests.cy.js:161]: The wording for these tests are a bit confusing when reading the output. 

`Cancel on ESC Keypress ESC key cancels any changes`
`Cancel on ESC Keypress Enter key changes textfield value`

Is there a way to rename them so they are a bit more descriptive of what is being tested? I think its mostly the `describe` that makes it confusing.

**Review by tuckerzp — CHANGES_REQUESTED** [review]
I have a few comments on the naming of the tests. I think it is important to be able to understand the intention of the test without needing to look at code for the output logs.

File: end-to-end-tests/cypress/e2e/OSCAL/OSCAL_Edit_Tests.cy.js (hunk 3 above)
Comment (schen2166) [reply end-to-end-tests/cypress/e2e/OSCAL/OSCAL_Edit_Tests.cy.js:161]: Sure thing, I can rename the tests. I was unsure if you have already renamed them in your request to change the test names in your other pull request. I can scrap the names have at the moment and reword them and make them more clear what the tests are doing.

Are the `Enter` and `Esc` test suites in the other pull request which your edit the test titles?

Comment (tuckerzp) [reply end-to-end-tests/cypress/e2e/OSCAL/OSCAL_Edit_Tests.cy.js:161]: If you look at #111, I have updated many of the tests made before your changes. As your `Editable Textfield` tests are not currently in `main`, I did not touch them

**Review by tuckerzp — APPROVED** [review]
Thanks for your work on this @schen2166! I am happy it will finally be merged in.

**Review by laurelmay — APPROVED** [review]
Thanks for these! Looks pretty good!

**Pull request merged**